SILCombine: Constant-fold MemoryLayout<T>.offset(of: \.literalKeyPath)

eeckstein opened the pull request:
Replace a call of the getter of AnyKeyPath._storedInlineOffset with a "constant" offset, in case of a keypath literal.
"Constant" offset means a series of struct_element_addr and tuple_element_addr instructions with a 0-pointer as base address.
These instructions can then be lowered to "real" constants in IRGen for concrete types, or to metatype offset lookups for generic or resilient types.

Replace:
```
  %kp = keypath ...
  %offset = apply %_storedInlineOffset_method(%kp)
```
with:
```
  %zero = integer_literal $Builtin.Word, 0
  %null_ptr = unchecked_trivial_bit_cast %zero to $Builtin.RawPointer
  %null_addr = pointer_to_address %null_ptr
  %projected_addr = struct_element_addr %null_addr
   ... // other address projections
  %offset_ptr = address_to_pointer %projected_addr
  %offset_builtin_int = unchecked_trivial_bit_cast %offset_ptr
  %offset_int = struct $Int (%offset_builtin_int)
  %offset = enum $Optional<Int>, #Optional.some!enumelt, %offset_int
```

rdar://problem/53309403

eeckstein committed:
SILCombine: Constant-fold MemoryLayout<T>.offset(of: \.literalKeyPath)

Replace a call of the getter of AnyKeyPath._storedInlineOffset with a "constant" offset, in case of a keypath literal.
"Constant" offset means a series of struct_element_addr and tuple_element_addr instructions with a 0-pointer as base address.
These instructions can then be lowered to "real" constants in IRGen for concrete types, or to metatype offset lookups for generic or resilient types.

Replace:
  %kp = keypath ...
  %offset = apply %_storedInlineOffset_method(%kp)
with:
  %zero = integer_literal $Builtin.Word, 0
  %null_ptr = unchecked_trivial_bit_cast %zero to $Builtin.RawPointer
  %null_addr = pointer_to_address %null_ptr
  %projected_addr = struct_element_addr %null_addr
   ... // other address projections
  %offset_ptr = address_to_pointer %projected_addr
  %offset_builtin_int = unchecked_trivial_bit_cast %offset_ptr
  %offset_int = struct $Int (%offset_builtin_int)
  %offset = enum $Optional<Int>, #Optional.some!enumelt, %offset_int

rdar://problem/53309403

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci benchmark

eeckstein commented:
@swift-ci smoke test linux

jckarter commented:
This is clever, but it relies on undefined behavior. 0 is not a valid value for a `Builtin.RawPointer`. I was thinking we could do this as an IRGen lowering pass, which would let us directly replace the call with the offset computation emitted by IRGen.

jckarter commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -271,6 +271,120 @@ bool SILCombiner::tryOptimizeKeypathApplication(ApplyInst *AI,
+    case KeyPathPatternComponent::Kind::OptionalChain:
+    case KeyPathPatternComponent::Kind::OptionalForce:
+    case KeyPathPatternComponent::Kind::OptionalWrap:
+      hasOffset = false;
```
We should not try to optimize if the key path component refers to a resilient external property, because the property being referenced may in fact be stored. If `component.getExternalDecl()` returns nonnull, we should leave the call in place.

eeckstein replied:
That's a good point.

eeckstein commented:
@jckarter Hm, I would prefer to keep that as a SIL optimization. How about using a Builtin.baseAddressForOffset instead of the 0-rawpointer (which is lowered to a null pointer in IRGen)?

jckarter commented:
@eeckstein That could work.

eeckstein committed:
SIL: a new instruction 'base_addr_for_offset' for field offset calculations.

The ``base_addr_for_offset`` instruction creates a base address for offset calculations.
The result can be used by address projections, like ``struct_element_addr``, which themselves return the offset of the projected fields.
IR generation simply creates a null pointer for ``base_addr_for_offset``.

eeckstein commented:
@jckarter I pushed a new version
(I created a new instruction rather than a builtin, because IMO it makes more sense).

eeckstein commented:
@swift-ci smoke test

jckarter reviewed (approved):
Looks good, thanks!
